File: src/main/scala/booklet/pure/http/Query.scala
package booklet.pure.http

import booklet.pure.Failure
import zhttp.http.Request
import zio.IO

object Query {

  def param(request: Request)(name: String): Option[String] =
    request.url.queryParams.get(name).flatMap(_.headOption)

  def fromRequest(request: Request): IO[Failure, Map[String, String]] =
    request.bodyAsString.mapBoth(Failure.fromThrowable, fromFormBody)

  def fromFormBody(body: String): Map[String, String] =
    body
      .split("\n")
      .filterNot(_.startsWith("----------------------------"))
      .grouped(3)
      .map { field =>
        val name = field(0).trim
          .stripPrefix("Content-Disposition: form-data; name=\"")
          .stripSuffix("\"")
        val value = field(2).trim
        name -> value
      }
      .toMap
}

File: src/main/ts/booklet/pure/http/Query.ts
export class Query {
  static fromRequest(request: Request): Map<string, string> {
    const url = new URL(request.url);
    const params = new URLSearchParams(url.search);
    const queryMap = new Map<string, string>();

    params.forEach((value, key) => {
      queryMap.set(key, value);
    });

    return queryMap;
  }
}
